### ISSUE TITLE
Jsonnet libraries for Terraform

### ISSUE THREAD

**Issue by sc250024** [hashicorp/terraform#23592]
### Description

Recently I've discovered that Terraform configurations can be written in JSON using the extension `tf.json`. I found this out after reading the following article:

* [Declarative Infrastructure with the Jsonnet Templating Language](https://databricks.com/blog/2017/06/26/declarative-infrastructure-jsonnet-templating-language.html)

This is very useful because I'm already using a Terragrunt-based repository, using modules as often as possible, to keep things as DRY as possible. However, there are always some cases where writing a module to do something simple is overkill. Using Jsonnet as a basis for writing the Terraform configurations is great because I can have the best of both worlds. Namely, I can use modules for the things I do most often, but I can also use Jsonnet to template the miscellaneous.

However, I found that often I am writing Jsonnet libraries (`.libsonnet` files) that mirror what Terraform itself is doing. As a very simple example, here's a function for [`data.aws_caller_identity`](https://www.terraform.io/docs/providers/aws/d/caller_identity.html), written in a form that's usable in Jsonnet:

```jsonnet
{
  /* Function to obtain AWS caller identity */
  aws_caller_identity(alias='current', provider=''):: {
    local current = if provider != '' then { provider: provider } else {},
    aws_caller_identity: {
      [alias]: current,
    },
  },
}
```

While this does the job, it would be nice not to have to write validation for every single Terraform function since Terraform itself (or the providers, specifically) already have this information.

Is it technically possible to take the resource specs written in Go, and generate Jsonnet libraries from them? I know that projects like Grafana maintain official libraries to help with creating complex configurations (like [grafana/grafonnet-lib](https://github.com/grafana/grafonnet-lib) for writing Grafana dashboards), but I'm not sure what's required to do this for a project like Terraform.

### Current Terraform Version

**Terraform v0.12.17**

### Use-cases

Numerous! Essentially it would give a rich new methodology for generating Terraform configurations. However, I'm not sure how the Terraform maintainers / Hashicorp bigwigs feel about people **not** using HCL.

### Attempted Solutions
 
As mentioned above, currently anytime I want to template a function for Terraform in Jsonnet, I'm more or less looking at the Terraform documentation for a provider (AWS most often), and implementing the required / optional values manually.

### Proposal

I'm a bit of a Golang newbie, but from what I can tell, the providers use the library `github.com/hashicorp/terraform-plugin-sdk/helper/schema` to create `Resource` objects which contain the definitions I'm talking about. Here's an example for getting an AMI as a data source: https://github.com/terraform-providers/terraform-provider-aws/blob/master/aws/data_source_aws_ami.go#L19-L183

Basically, my idea would be to make the SDK capable of also generating Jsonnet libraries.

### References

N/A

**Comment by teamterraform** [comment]
Hi @sc250024!

You can run [`terraform providers schema -json`](https://www.terraform.io/docs/commands/providers/schema.html) to get a JSON-serialized schema for all of the providers in the current configuration. That is the public interface for accessing schema information, rather than calling directly into the providers as if they were go libraries. (Providers are Go _applications_, not libraries.)

If you want to get the schema for a particular provider without adding it to your main configuration, you can get there by writing a simple placeholder configuration that just includes the dependencies on the provider(s) you are interested in:

```hcl
terraform {
  required_providers {
    aws = "2.7.0"
  }
}
```

If you run `terraform init` in the directory containing the above file then Terraform will install the requested AWS provider version such that when you run `terraform providers schema -json` the result will include the AWS provider schema. (That provider's schema is a particularly large one, so the result may be overwhelming to print to your terminal.)

While we don't intend to maintain Jsonnet libraries for Terraform providers ourselves -- we already have our own language to maintain, after all -- but if you're interested in working on such a thing we suspect you could get the information you need from the above command and translate that result into Jsonnet syntax in a way that suits your needs.

As for how we feel about folks not using the HCL native syntax: we provide the JSON variant of the Terraform language to allow exactly this sort of thing, and if Jsonnet helps you solve your problems then please do use it! Do note though that Terraform has its own expression language, even in the JSON variant, and so it's likely that you'll need to use Jsonnet expressions to construct strings containing Terraform expressions, which may cause your input files to be confusing to read.

Since direct Jsonnet support is not something we'll be able to offer within Terraform itself, we're going to close this now. Hopefully the above information is useful for getting the data you need!
